File: tmp_puzzle_test.js
const types = ['fire','water','wind','thunder'];

function initPuzzleGrid() {
  const newGrid = [];
  for (let r = 0; r < 7; r++) {
    const row = [];
    for (let c = 0; c < 7; c++) {
      let t;
      do {
        t = types[Math.floor(Math.random() * types.length)];
      } while (
        (r >= 2 && newGrid[r-1][c].type === t && newGrid[r-2][c].type === t) ||
        (c >= 2 && row[c-1].type === t && row[c-2].type === t)
      );
      row.push({ id: Math.random(), type: t });
    }
    newGrid.push(row);
  }
  return newGrid;
}

function findMatches(grid) {
  const horizontalItems = new Map();
  const verticalItems = new Map();

  for (let r = 0; r < 7; r++) {
    let count = 1;
    let startC = 0;
    for (let c = 1; c <= 7; c++) {
      if (c < 7 && grid[r][c].type && grid[r][c].type === grid[r][c-1].type) {
        count++;
      } else {
        if (count >= 3) {
          const matchSet = new Set();
          for (let i = startC; i < c; i++) matchSet.add(`${r},${i}`);
          horizontalItems.set(`${r},${startC}-${count}`, matchSet);
        }
        startC = c;
        count = 1;
      }
    }
  }

  for (let c = 0; c < 7; c++) {
    let count = 1;
    let startR = 0;
    for (let r = 1; r <= 7; r++) {
      if (r < 7 && grid[r][c].type && grid[r][c].type === grid[r-1][c].type) {
        count++;
      } else {
        if (count >= 3) {
          const matchSet = new Set();
          for (let i = startR; i < r; i++) matchSet.add(`${i},${c}`);
          verticalItems.set(`${startR}-${count},${c}`, matchSet);
        }
        startR = r;
        count = 1;
      }
    }
  }

  const allMatches = [];
  horizontalItems.forEach((set, key) => {
    const [r, range] = key.split(',');
    const [start] = range.split('-').map(Number);
    const coords = Array.from(set).map(s => s.split(',').map(Number));
    allMatches.push({ coords, type: grid[Number(r)][start].type, direction: 'h' });
  });
  verticalItems.forEach((set, key) => {
    const [range, c] = key.split(',');
    const [start] = range.split('-').map(Number);
    const coords = Array.from(set).map(s => s.split(',').map(Number));
    allMatches.push({ coords, type: grid[start][Number(c)].type, direction: 'v' });
  });

  return allMatches;
}

for (let i = 0; i < 20; i++) {
  const grid = initPuzzleGrid();
  const matches = findMatches(grid);
  if (matches.length !== 0) {
    console.log('FOUND ERROR', i, matches);
    process.exit(1);
  }
}
console.log('no initial matches');

File: tmp_puzzle_test.py
import random

types = ['fire','water','wind','thunder']

def initPuzzleGrid():
    newGrid = []
    for r in range(7):
        row = []
        for c in range(7):
            while True:
                t = random.choice(types)
                if r >= 2 and newGrid[r-1][c]['type'] == t and newGrid[r-2][c]['type'] == t:
                    continue
                if c >= 2 and row[c-1]['type'] == t and row[c-2]['type'] == t:
                    continue
                break
            row.append({'id': random.random(), 'type': t})
        newGrid.append(row)
    return newGrid


def findMatches(grid):
    horizontalItems = {}
    verticalItems = {}

    for r in range(7):
        count = 1
        startC = 0
        for c in range(1, 8):
            if c < 7 and grid[r][c]['type'] and grid[r][c]['type'] == grid[r][c-1]['type']:
                count += 1
            else:
                if count >= 3:
                    matchSet = set(f"{r},{i}" for i in range(startC, c))
                    horizontalItems[f"{r},{startC}-{count}"] = matchSet
                startC = c
                count = 1

    for c in range(7):
        count = 1
        startR = 0
        for r in range(1, 8):
            if r < 7 and grid[r][c]['type'] and grid[r][c]['type'] == grid[r-1][c]['type']:
                count += 1
            else:
                if count >= 3:
                    matchSet = set(f"{i},{c}" for i in range(startR, r))
                    verticalItems[f"{startR}-{count},{c}"] = matchSet
                startR = r
                count = 1

    allMatches = []
    for key, setv in horizontalItems.items():
        r, range_ = key.split(',')
        start, _ = map(int, range_.split('-'))
        coords = [tuple(map(int, s.split(','))) for s in setv]
        allMatches.append({'coords': coords, 'type': grid[int(r)][start]['type'], 'direction': 'h'})

    for key, setv in verticalItems.items():
        range_, c = key.split(',')
        start, _ = map(int, range_.split('-'))
        coords = [tuple(map(int, s.split(','))) for s in setv]
        allMatches.append({'coords': coords, 'type': grid[start][int(c)]['type'], 'direction': 'v'})

    return allMatches

for i in range(20):
    grid = initPuzzleGrid()
    matches = findMatches(grid)
    if matches:
        print('FOUND ERROR', i, matches)
        break
else:
    print('no initial matches')
